python312Packages.openusd: fix build

ShaddyDC opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Changelog: https://github.com/PixarAnimationStudios/OpenUSD/blob/59992d2178afcebd89273759f2bddfe730e59aa8/CHANGELOG.md

This is not the latest version as the patch used isn't updated to that version yet.

I tested the basic functionality of _some_ of the binaries.

6 packages built:
blender blender-hip blendfarm openusd python311Packages.openusd python312Packages.openusd

fixes https://github.com/NixOS/nixpkgs/issues/352873

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ShaddyDC committed:
python312Packages.openusd: apply upstream patch to fix build

Co-authored-by: ShadyDC <shaddythefirst@gmail.com>
Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4780

dotlambda commented on pkgs/development/python-modules/openusd/default.nix:
```diff
@@ -51,13 +51,13 @@ in
 buildPythonPackage rec {
   pname = "openusd";
-  version = "24.05";
+  version = "24.08";
```
```suggestion
  version = "24.08";
  pyproject = false;
```
instead of `format`

dotlambda reviewed (commented):
Please rebase.
The commit message should start with `python312Packages.` and not talk about a version bump because 24.08 is already in master.

nyabinary commented:
In the changelog
"OpenUSD is now licensed under the Tomorrow Open Source Technology license. This is a renaming of the license, but the terms and conditions are the same as before. See [this forum post](https://forum.aousd.org/t/upcoming-openusd-license-update/1561) for details." 

So the license needs to be updated too.

gador committed:
lib/licenses.nix: add tost

this adds the license for OpenUSD which is explained
here https://forum.aousd.org/t/upcoming-openusd-license-update/1561

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented:
This fails to build with:
```
openusd> -- Configuring done (4.6s)
openusd> CMake Error at /nix/store/v6zrwwq9l9zdd3bdz6mqwvasc77mjkz8-python3.11-materialx-1.39.1/lib/cmake/MaterialX/MaterialXTargets.cmake:222 (set_target_properties):
openusd>   The link interface of target "MaterialXRender" contains:
openusd>
openusd>     OpenImageIO::OpenImageIO
openusd>
openusd>   but the target was not found.  Possible reasons include:
openusd>
openusd>     * There is a typo in the target name.
openusd>     * A find_package call is missing for an IMPORTED target.
openusd>     * An ALIAS target is missing.
openusd>
openusd> Call Stack (most recent call first):
openusd>   /nix/store/v6zrwwq9l9zdd3bdz6mqwvasc77mjkz8-python3.11-materialx-1.39.1/lib/cmake/MaterialX/MaterialXConfig.cmake:36 (include)
openusd>   cmake/defaults/Packages.cmake:315 (find_package)
openusd>   CMakeLists.txt:23 (include)
openusd>
openusd>
openusd> -- Generating done (0.9s)
openusd> CMake Warning:
openusd>   Manually-specified variables were not used by the project:
openusd>
openusd>     BUILD_TESTING
openusd>     CMAKE_INSTALL_BINDIR
openusd>     CMAKE_INSTALL_DOCDIR
openusd>     CMAKE_INSTALL_INCLUDEDIR
openusd>     CMAKE_INSTALL_INFODIR
openusd>     CMAKE_INSTALL_LIBDIR
openusd>     CMAKE_INSTALL_LIBEXECDIR
openusd>     CMAKE_INSTALL_LOCALEDIR
openusd>     CMAKE_INSTALL_MANDIR
openusd>     CMAKE_INSTALL_OLDINCLUDEDIR
openusd>     CMAKE_INSTALL_SBINDIR
openusd>     CMAKE_POLICY_DEFAULT_CMP0025
openusd>     PYSIDE_BIN_DIR
openusd>
openusd>
openusd> CMake Generate step failed.  Build files cannot be regenerated correctly.
```

gador commented:
ok, this error is due the update of `materialX` in 993ec329851b363c24ceac95034c6bc7111acb0b by @mweinelt 
As I've said in https://github.com/NixOS/nixpkgs/pull/326466#issuecomment-2293033805 `materialX` should not have been updated.
It may have been a good idea to go through a PR workflow  and have me tagged in it. I then would have been able to catch that.

The only reasonable way forward in light of ZHF #352882 is to revert the update from the above commit.

I tested this PR with it and it builds fine, then.

gador committed:
Revert "python3Packages.materialx: 1.38.10 -> 1.39.1"

This reverts commit 993ec329851b363c24ceac95034c6bc7111acb0b.
MaterialX has some upstream issues which haven't been quite
worked out, yet[1]. It is therefore reasonable to wait
with an update in nixpkgs. It also breaks downstream packages
like python312Packages.openusd as well as blender[2]

[1]: https://github.com/PixarAnimationStudios/OpenUSD/pull/3159
[2]: https://github.com/NixOS/nixpkgs/pull/326466#issuecomment-2293033805

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
python312Packages.openusd: apply upsteam's new license

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented:
@ShaddyDC I piggybacked your PR here, did the rebase and added the new license. I hope thats ok.
@mweinelt could you have a look at this PR, as it also contains the revert of the material update and let me know, if you're ok with it?

mweinelt commented:
The revert should contain the reason it was reverted.

amarshall reviewed (commented):
- PR title needs to be updated, it doesn’t reflect what’s happening here anymore
- Ideally the materialx revert should be first, otherwise the patch fix doesn’t actually fix the build

gador committed:
python312Packages.openusd: conform to PEP 517

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented on pkgs/development/python-modules/openusd/default.nix:
```diff
@@ -52,6 +52,7 @@ in
 buildPythonPackage rec {
   pname = "openusd";
   version = "24.08";
+  pyproject = false;
```
done

gador commented on pkgs/development/python-modules/openusd/default.nix:
```diff
@@ -65,13 +66,11 @@ buildPythonPackage rec {
       name = "port-to-embree-4.patch";
-      url = "https://github.com/PixarAnimationStudios/OpenUSD/pull/2266/commits/4b6c23d459c602fdac5e0ebc9b7722cbd5475e86.patch";
-      hash = "sha256-yjqdGAVqfEsOX1W/tG6c+GgQLYya5U9xgUe/sNIuDbw=";
+      url = "https://github.com/PixarAnimationStudios/OpenUSD/pull/2266/commits/c8fec1342e05dca98a1afd4ea93c7a5f0b41e25b.patch";
```
done

Atemu replied:
@amarshall this is the first I hear of this parameter. Could you elaborate? Is this something we always want? If so, we should make it conventional or even required (via RFC) and perhaps make `fetchpatch` apply it automatically.

Scrumplex replied:
`full_index=1` uses full commit hashes instead of shortened ones. I don't think it matters too much in this case, as we aren't applying patches using `git apply`/`git am`

ShaddyDC commented:
@gador thanks for taking this over! Those changes look good to me!

Scrumplex commented on pkgs/development/python-modules/materialx/default.nix:
```diff
@@ -17,13 +17,13 @@
     repo = "MaterialX";
-    rev = "refs/tags/v${version}";
-    hash = "sha256-WzzsY1hOWwJEqT/ZRLIoZDfKNvx1Yf6aFhA3ZcSPx+s=";
+    rev = "v${version}";
```
Why was `refs/tags` removed?

```suggestion
    rev = "refs/tags/v${version}";
```

gador replied:
Because this is a direct revert of the update commit

Atemu replied:
Ah, that explains it. Do re-revert this though, it's okay if a revert isn't an exact revert.

Atemu commented on lib/licenses.nix:
```diff
@@ -1212,6 +1212,12 @@ lib.mapAttrs mkLicense ({
     fullName = "TCL/TK License";
   };
 
+  tost = {
```
Where does this name come from?

If you made it up, I think a direct acronym would be better.

```suggestion
  tostl = {
```

Atemu commented on lib/licenses.nix:
```diff
@@ -1212,6 +1212,12 @@ lib.mapAttrs mkLicense ({
   };
 
+  tost = {
+    shortName = "TSL";
```
Is this a short name for which there is precedent? If not, I'd rather just not declare it.

  ```suggestion
```

gador replied:
yes, also from https://forum.aousd.org/t/upcoming-openusd-license-update/1561

Atemu replied:
Those 3 letters do not appear on that page in that order.

gador replied:
Oh, yes I see that now. This is a typo I think and should either be "TOST" or removed

Atemu replied:
This defaults to the license attr name, so simply remove it.

Scrumplex commented on pkgs/development/python-modules/openusd/default.nix:
```diff
@@ -65,13 +66,12 @@ buildPythonPackage rec {
-      url = "https://github.com/PixarAnimationStudios/OpenUSD/pull/2266/commits/4b6c23d459c602fdac5e0ebc9b7722cbd5475e86.patch";
-      hash = "sha256-yjqdGAVqfEsOX1W/tG6c+GgQLYya5U9xgUe/sNIuDbw=";
+      # https://github.com/PixarAnimationStudios/OpenUSD/pull/2266
+      url = "https://github.com/PixarAnimationStudios/OpenUSD/commit/c8fec1342e05dca98a1afd4ea93c7a5f0b41e25b.patch?full_index=1";
```
Is `full_index=1` needed?

nyabinary replied:
https://github.com/NixOS/nixpkgs/pull/351902#discussion_r1827042323
Is why `full_index=1` was added

Scrumplex replied:
It seems like `full_index=1` just expands the commit hashes used in the diff. I don't think it's strictly needed, but at the same time it doesn't really hurt.

Atemu commented on lib/licenses.nix:
```diff
@@ -1212,6 +1212,12 @@ lib.mapAttrs mkLicense ({
 
+  tost = {
+    shortName = "TSL";
+    fullName = "Tomorrow Open Source License";
```
```suggestion
    fullName = "Tomorrow Open Source Technology License 1.0";
```

gador replied:
Thats not the official name

Atemu replied:
It's what it says right in the license text.

gador replied:
> To date, USD has been distributed under what we called a modified Apache 2.0 License. We now have a new name for our license: the “Tomorrow Open Source License” or “TOST.”

as it is written on there blog. But I agree with you that we should name it after their official LICENSE text

Atemu replied:
It's also good to include the version at least here in the fullName as they may have a v2 at some point.

Atemu commented on lib/licenses.nix:
```diff
@@ -1212,6 +1212,12 @@ lib.mapAttrs mkLicense ({
+  tost = {
+    shortName = "TSL";
+    fullName = "Tomorrow Open Source License";
+    url = "https://github.com/PixarAnimationStudios/OpenUSD/blob/release/LICENSE.txt";
```
If this is the only project which uses this license (as evidenced by the fact that it's only available in their source repo), I think it'd be better to just have it as a custom license in `meta.license` without adding it to `lib.licenses`.

nyabinary replied:
> If this is the only project which uses this license (as evidenced by the fact that it's only available in their source repo), I think it'd be better to just have it as a custom license in `meta.license` without adding it to `lib.licenses`.

TBF they could use it for future projects as well (Pixar, I mean), but yeah.

Atemu replied:
Good to know, in that case it should be added and used in both packages.

Atemu commented on pkgs/development/python-modules/openusd/default.nix:
```diff
@@ -175,7 +175,7 @@ buildPythonPackage rec {
     homepage = "https://openusd.org/";
-    license = lib.licenses.asl20;
+    license = lib.licenses.tost;
     maintainers = with lib.maintainers; [ shaddydc ];
```
Given that you're familiar with the drv, you might want to add yourself as maintainer @gador

gador replied:
Yes, maybe not a bad idea..

Atemu reviewed (approved):
Diff LGTM

gador commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

## `nixpkgs-review` result
<details>
<summary>
:fast_forward: 1 package blacklisted:
</summary>
<ul>
<li>nixos-install-tools</li>
</ul>
</details>
<details>
<summary>
:white_check_mark: 9 packages built:
</summary>
<ul>
<li>blender</li>
<li>blender-hip</li>
<li>blendfarm</li>
<li>materialx (python312Packages.materialx)</li>
<li>nixpkgs-manual</li>
<li>openusd</li>
<li>python311Packages.materialx</li>
<li>python311Packages.openusd</li>
<li>python312Packages.openusd</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
blender-hip-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘blender-hip-x86_64-linux’ attribute.
</details>
blender-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘blender-x86_64-linux’ attribute.
</details>
blendfarm-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘blendfarm-x86_64-linux’ attribute.
</details>
materialx-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘materialx-x86_64-linux’ attribute.
</details>
nixpkgs-manual-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘nixpkgs-manual-x86_64-linux’ attribute.
</details>
openusd-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘openusd-x86_64-linux’ attribute.
</details>
python311Packages.materialx-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘python311Packages.materialx-x86_64-linux’ attribute.
</details>
python311Packages.openusd-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘python311Packages.openusd-x86_64-linux’ attribute.
</details>
python312Packages.openusd-x86_64-linux:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/1j71lcyxcsn2pj4ajw7rskj15pa775vy-nixpkgs’ do not contain ‘python312Packages.openusd-x86_64-linux’ attribute.
</details>
</details>

gador commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

## `nixpkgs-review` result
<details>
<summary>
:fast_forward: 1 package marked as broken and skipped:
</summary>
<ul>
<li>blender</li>
</ul>
</details>
<details>
<summary>
:white_check_mark: 6 packages built:
</summary>
<ul>
<li>materialx (python312Packages.materialx)</li>
<li>nixpkgs-manual</li>
<li>openusd</li>
<li>python311Packages.materialx</li>
<li>python311Packages.openusd</li>
<li>python312Packages.openusd</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
materialx-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘materialx-aarch64-darwin’ attribute.
</details>
nixpkgs-manual-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘nixpkgs-manual-aarch64-darwin’ attribute.
</details>
openusd-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘openusd-aarch64-darwin’ attribute.
</details>
python311Packages.materialx-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘python311Packages.materialx-aarch64-darwin’ attribute.
</details>
python311Packages.openusd-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘python311Packages.openusd-aarch64-darwin’ attribute.
</details>
python312Packages.openusd-aarch64-darwin:
<details>

Package is missing a meta section.
If the package is using runCommand please make sure to inherit or create a meta section.

error: AttrPathNotFound
Packages in ‘/nix/store/sj0krjxk6f39ys1znksm3cxhlk4q1d3j-nixpkgs’ do not contain ‘python312Packages.openusd-aarch64-darwin’ attribute.
</details>
</details>

gador committed:
python312Packages.openusd: add gador as co-maintainer

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

philiptaron reviewed (approved):
Good by me.
